service/resource/resourcegroup/resource.go

```diff
@@ -0,0 +1,298 @@
+
+// GetUpdateState returns nil for the create, delete and update states because
+// resource groups are not updated.
+func (r *Resource) GetUpdateState(obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
+	return nil, nil, nil, nil
+}
+
+// Name returns the resource name.
+func (r *Resource) Name() string {
+	return Name
+}
+
+// ProcessCreateState creates the resource group via the Azure API.
+func (r *Resource) ProcessCreateState(obj, createState interface{}) error {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	resourceGroupToCreate, err := toResourceGroup(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if resourceGroupToCreate != nil {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating the resource group in the Azure API")
+
+		groupClient, err := r.getGroupsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		_, err = groupClient.CreateOrUpdate(*resourceGroupToCreate.Name, *resourceGroupToCreate)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "created the resource group in the Azure API")
+
+	} else {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "the resource group already exists in the Azure API")
+	}
+
+	return nil
+}
+
+// ProcessDeleteState deletes the resource group via the Azure API.
+func (r *Resource) ProcessDeleteState(obj, deleteState interface{}) error {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	resourceGroupToDelete, err := toResourceGroup(deleteState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if resourceGroupToDelete != nil {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "deleting the resource group in the Azure API")
+
+		groupsClient, err := r.getGroupsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Delete the resource group which also deletes all resources it
+		// contains. We wait for the error channel while the deletion happens.
+		_, errchan := groupsClient.Delete(*resourceGroupToDelete.Name, nil)
+		err = <-errchan
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "deleted the resource group in the Azure API")
+
+	} else {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "the resource group does not exist in the Azure API")
+	}
+
+	return nil
+}
+
+// ProcessUpdateState returns nil because resource groups are not updated.
+func (r *Resource) ProcessUpdateState(obj, updateState interface{}) error {
+	return nil
+}
+
+// Underlying returns the underlying resource.
+func (r *Resource) Underlying() framework.Resource {
+	return r
+}
+
+func (r *Resource) getGroupsClient() (*azureresource.GroupsClient, error) {
+	azureClients, err := client.NewAzureClientSet(r.azureConfig)
```

Comment:
Currently, we create new API clients per request which is wasteful. #9 tracks fixing this.